### ISSUE TITLE
Add a conditional branching combinator

### ISSUE THREAD

**Issue by Gozala** [Gozala/decoder.flow#2]
Elm provides `andThen : (a -> Decoder b) -> Decoder a -> Decoder b`  function to create decoders that depend on previous results. For example if you are creating versioned data, you might do something like this:

```elm
info : Decoder Info
info =
  field "version" int
    |> andThen infoHelp

infoHelp : Int -> Decoder Info
infoHelp version =
  case version of
    4 ->
      infoDecoder4

    3 ->
      infoDecoder3

    _ ->
      fail <|
        "Trying to decode info, but version "
        ++ toString version ++ " is not supported."

-- infoDecoder4 : Decoder Info
-- infoDecoder3 : Decoder Info
```

decoder.flow does not provide `andThen` as it would make serialization and transfer of decoders across the threads impossible. Although similar to how `Decoder.record` and `Decoder.form` provide a way to address same use cases as `map`, `map2`, ...`map8` in Elm  we could provide some solution to addressing `andThen` use cases like the one above. For instance `match` like combinator could be implemented:

```js
const version = Decoder.field("version", Decoder.Integer)
Decoder.either(
  Decoder.when(version, Decoder.ok(4), infoDecoder4)
  Decoder.when(version, Decoder.ok(3), infoDecoder3)
  Decoder.error("Trying to decode info, but provided version isn't supported"))
```

Note that in comparison to `andThen` this is far more limited and even this example unlike original Elm code is unable to include encountered version in the error message, but it still might enable certain use cases that aren't possible today.

Primary limitation of this would be that unlike `andThen` result can't be carried over to the next decoder, but maybe that could be worked around like in the example below:

```js
const versionedInfo = Decoder.form({
  version: Decoder.field("version", Decoder.Integer)
  info: Decoder.value
})

const infoHelp = Decoder.either(
  Decoder.record({ version: Decoder.ok(4), ok: infoDecoder4 }),
  Decoder.record({ version: Decoder.ok(3), ok: infoDecoder3 }),
  Decoder.record({ version: Decoder.Integer, error: Decoder.ok('Invalid version') }))

const info = Decoder.chain(versionedInfo, infoHelp)
```

P.S.: We currently have no `Decoder.value` nor `Decoder.chain` but they could be easily added.

**Comment by asajeffrey** [comment]
Would it good enough to support higher-order combinators for decoders where the functions are required to be top-level exports from a Node module? Then they could be serialized between threads using the path to the function. It would support a form of map and co, and would probably do this example too.
`

**Comment by Gozala** [comment]
> Would it good enough to support higher-order combinators for decoders where the functions are required to be top-level exports from a Node module?

Well those still may depend on local state, but even ignoring that how do we enforce that requirement, I think at best we could just ask nicely.

**Comment by asajeffrey** [comment]
True, there's no much we can do if people use global variables, but I think getting both threads to load modules, and then using the module export name would at least give something we can do. Worklets use a similar trick of saying in the spec that there may be more than one global.

**Comment by Gozala** [comment]
Think more about this I'm leaning towards adding following two combinators:

```js
and(Decoder<a>, Decoder<b>):Decoder<b>
```
Takes two decoders, when both succeed returns result of the second decoder, if either fails returns error of first failure.

```js
match <a:JSON>(value:a):Decoder<a>
```

Takes JSON value and return decoder that succeeds with given value if input matches it or fails otherwise.

With that it would also make sense to rename or alias `Decoder.either` as `Decoder.or`. With this combinators in place original example could be implemented as follows:

```js
const matchVersion = (version) =>
  Decoder.field("version", Decoder.match(version))

const info = Decoder.or(
  Decoder.and(matchVersion(4), infoDecoder4),
  Decoder.and(matchVersion(3), infoDecoder3)
)
```

P.S. Note that we did not include `Decoder.error` at the end as `Decoder.or` / `Decoder.either` will include detailed error message stating both encountered version numbers and expected version numbers.

**Comment by sshquack** [comment]
Thanks for publishing 0.5 with `match` and `or` support! I assume this issue can be closed. Looking forward to using it with `ChainDecoder` or `ExternalDecoder`
